File: EditorCompresor/cpp/editor.cpp
#include <ncurses.h> //sudo apt-get install libncurses5-dev libncursesw5-dev
#include "editor.h"

void guardar_comprimido(const std::string& buffer, const std::string& archivo) {
    std::vector<int> comprimido = comprimir(buffer);
    std::ofstream output(archivo, std::ios::binary);
    for (int numero : comprimido) {
        output.put(char((numero >> 8) & 0xFF));
        output.put(char(numero & 0xFF));
    }
    output.close();
}

std::string leer_comprimido(const std::string& archivo) {
    std::ifstream input(archivo, std::ios::binary);
    std::vector<int> comprimido;
    char bytes[2];
    while (input.read(bytes, 2)) {
        int numero = (unsigned char)(bytes[0]) << 8 | (unsigned char)(bytes[1]);
        comprimido.push_back(numero);
    }
    input.close();
    return descomprimir(comprimido);
}

void editar_buffer(WINDOW* win, std::string& buffer) {
    int ch;
    int row = 0;
    int col = 0;
    bool running = true;

    while (running) {
        wclear(win);
        mvwprintw(win, 0, 0, buffer.c_str());  // Imprime el buffer actual
        wmove(win, row, col);  // Mueve el cursor a la posición actual

        ch = wgetch(win);
        switch (ch) {
            case 24:  // Ctrl-X para salir
                running = false;
                break;
            case 19:  // Ctrl-S para guardar
                {
                    char archivo[80];
                    echo();
                    mvwgetstr(win, 1, 0, archivo);
                    noecho();
                    guardar_comprimido(buffer, archivo);
                }
                break;
            case KEY_BACKSPACE:
            case 127:
                if (col > 0) {
                    buffer.erase(col - 1, 1);
                    col--;
                }
                break;
            case '\n':
                buffer.insert(col, 1, '\n');
                col = 0;
                row++;
                break;
            default:
                buffer.insert(buffer.begin() + col, ch);
                col++;
                break;
        }
    }
}

int main() {
    std::string buffer;
    char opcion;

    std::cout << "¿Desea abrir un archivo existente (S/N)? ";
    std::cin >> opcion;

    if (opcion == 'S' || opcion == 's') {
        char archivo[80];
        std::cout << "Nombre del archivo a abrir: ";
        std::cin >> archivo;
        buffer = leer_comprimido(archivo);
    }

    initscr();
    cbreak();
    noecho();
    keypad(stdscr, TRUE);

    editar_buffer(stdscr, buffer);

    endwin();
    return 0;
}

File: EditorCompresor/python/editor.py
from compresor import comprimir, guardar_comprimido
from descompresor import leer_comprimido, descomprimir


def editar_buffer(buffer):
    lines = buffer.split("\n")
    cursorRow = 0
    cursorCol = 0
    running = True

    while running:
        for i, line in enumerate(lines):
            if i == cursorRow:
                print("> ", end="")
            else:
                print("  ", end="")
            print(line)

        input_usuario = input()
        print("Comandos: Ctrl+S (Guardar), Ctrl+X (Salir)")

        if input_usuario == "\x1bS":  # Ctrl+S para guardar
            archivo = input("Nombre del archivo a guardar: ")
            guardar_comprimido(buffer, archivo)
        elif input_usuario == "\x1bX":  # Ctrl+X para salir
            running = False
        else:
            print("Input no reconocido. Use Ctrl+S para guardar y Ctrl+X para salir.")

    buffer = "\n".join(lines)


if __name__ == "__main__":
    buffer = ""
    opcion = input("¿Desea abrir un archivo existente (S/N)? ")
    if opcion.upper() == "S":
        archivo = input("Nombre del archivo a abrir: ")
        buffer = descomprimir(leer_comprimido(archivo))

    editar_buffer(buffer)
